### PR TITLE
[Distributed] Implement whenLocal properly

### PR DESCRIPTION
**Pull request by ktoso**
Resolves rdar://89082275

This aims to properly implement `whenLocal` so that it works properly in the type system, not by side stepping it. Almost there but need to consult a few pieces.

Replaces https://github.com/apple/swift/pull/42098

### CODE REVIEW

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -5635,9 +5638,13 @@ class ParamDecl : public VarDecl {
 
     /// Whether or not this parameter is '_const'.
     IsCompileTimeConst = 1 << 1,
+
+    // FIXME: it's not on the name but on the type
+    /// Whether or not this parameter is a '_local DistributedActor' reference.
+    IsDistributedKnownToBeLocal = 1 << 2,
```
Comment (ktoso) [inline include/swift/AST/Decl.h:5644]: not actually sure if we need it, we can ride on the attr

File: test/Distributed/Runtime/distributed_actor_whenLocal.swift [hunk 2]
```diff
@@ -30,18 +30,20 @@ func test() async throws {
 
   let local = Capybara(actorSystem: system)
   // await local.eat() // SHOULD ERROR
-  let valueWhenLocal: String? = await local.whenLocal { __secretlyKnownToBeLocal in
-    __secretlyKnownToBeLocal.eat()
+  let valueWhenLocal: String? = await local.whenLocal {
+    (definitelyLocal: _local Capybara) in await definitelyLocal.eat()
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_whenLocal.swift:34]: so explicitly annotating works, but it doesnt propagate implicitly if we just say `definitelyLocal in ...` 🤔

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (DougGregor) [reply include/swift/AST/Decl.h:5644]: It should be here, because we might end up inferring it in some cases (e.g., for a closure).

File: include/swift/AST/Types.h [hunk 3]
```diff
@@ -2023,20 +2024,24 @@ class ParameterTypeFlags {
...
       : value((variadic ? Variadic : 0) | (autoclosure ? AutoClosure : 0) |
               (nonEphemeral ? NonEphemeral : 0) |
               uint8_t(ownership) << OwnershipShift |
               (isolated ? Isolated : 0) |
               (noDerivative ? NoDerivative : 0) |
-              (compileTimeConst ? CompileTimeConst : 0)){}
+              (compileTimeConst ? CompileTimeConst : 0) |
+              (distributedKnownLocal ? DistributedKnownToBeLocal : 0)) {
+    if (distributedKnownLocal)
+      fprintf(stderr, "[%s:%d] (%s) distributedKnownLocal = %d\n", __FILE__, __LINE__, __FUNCTION__, distributedKnownLocal);
```
Comment (DougGregor) [inline include/swift/AST/Types.h:2036]: Stray `fprintf`

File: lib/AST/Expr.cpp [hunk 4]
```diff
@@ -277,6 +277,10 @@ bool Expr::isSemanticallyConstExpr(
   return printConstExprValue(nullptr, additionalCheck);
 }
 
+bool Expr::isDistributedKnownToBeLocal() const {
+  assert(false && "fixme: should know the attr");
```
Comment (DougGregor) [inline lib/AST/Expr.cpp:281]: I assume you'll be looking for a `DeclRefExpr` storing a `_local` parameter?

File: lib/Sema/CSApply.cpp [hunk 5]
```diff
@@ -7957,6 +7961,9 @@ bool ExprRewriter::requiresDistributedThunk(Expr *base, SourceLoc memberLoc,
     [&](ParamDecl *P) {
     return P->isIsolated() ||
            llvm::is_contained(solution.isolatedParams, P);
+  },
+    [&](ParamDecl *P) {
+    return P->isDistributedKnownToBeLocal(); // TODO: need the solution too?
```
Comment (DougGregor) [inline lib/Sema/CSApply.cpp:7966]: Yeah, you'll want to have the solution contain a set of known-to-be-local-parameters, like we did for isolated parameters, and do the `llvm::is_contained` check there.

File: lib/Sema/CSSimplify.cpp [hunk 6]
```diff
@@ -9864,13 +9864,23 @@ bool ConstraintSystem::resolveClosure(TypeVariableType *typeVar,
         auto flags = param.getParameterFlags();
 
         // Note when a parameter is inferred to be isolated.
-        if (contextualParam->isIsolated() && !flags.isIsolated() && paramDecl)
+        if (contextualParam->isIsolated() && !flags.isIsolated() && paramDecl) {
           isolatedParams.insert(paramDecl);
 
+          // an 'isolated' param is also implicitly '_local',
+          // since in order to run "on" an actor, it must be a real local instance.
+          distributedKnownLocalParams.insert(paramDecl);
```
Comment (DougGregor) [inline lib/Sema/CSSimplify.cpp:9872]: Hmm. This feels like a semantic property that should be checked elsewhere, e.g., a "known to be local" predicate that checks both `isolated` and `_local`, rather than inserting into both sets at this point. (Something we likely won't do consistently everywhere).

File: lib/Sema/CSSimplify.cpp [hunk 7]
```diff
@@ -9864,13 +9864,23 @@ bool ConstraintSystem::resolveClosure(TypeVariableType *typeVar,
...
+        } else if (contextualParam->isDistributedKnownToBeLocal() &&
```
Comment (DougGregor) [inline lib/Sema/CSSimplify.cpp:9873]: I think this should be an `if` not an `else if`.

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 8]
```diff
@@ -2133,14 +2133,21 @@ namespace {
         for (auto arg : *call->getArgs()) {
           if (auto declRef = dyn_cast<DeclRefExpr>(arg.getExpr())) {
             if (auto var = dyn_cast<VarDecl>(declRef->getDecl())) {
-              if (var->isSelfParameter()) {
-                return var;
-              }
+              return var;
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:2136]: This is somewhat of a surprise. Do we not need to check for `self` anywhere?

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 9]
```diff
@@ -2284,9 +2291,15 @@ namespace {
     /// context and false otherwise.
     bool isDistributedAccess(SourceLoc declLoc, ValueDecl *decl,
                              Expr *context) {
+      fprintf(stderr, "[%s:%d] (%s) decl:\n", __FILE__, __LINE__, __FUNCTION__);
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:2294]: Stray debugging output.

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 10]
```diff
@@ -2284,9 +2291,15 @@ namespace {
...
+      decl->dump();
+
       // If base of the call is 'local' we permit skip distributed checks.
-      if (auto baseSelf = findReferencedBaseSelf(context)) {
-        if (baseSelf->getAttrs().hasAttribute<KnownToBeLocalAttr>())
+      if (auto baseSelf = findReferencedCallBase(context)) {
+        fprintf(stderr, "[%s:%d] (%s) base:\n", __FILE__, __LINE__, __FUNCTION__);
+        baseSelf->dump();
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:2300]: More stray debugging output.

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 11]
```diff
@@ -2284,9 +2291,15 @@ namespace {
...
+
+        if (baseSelf->isDistributedKnownToBeLocal())
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:2302]: Oh, so this isn't necessarily `self`, but `isDistributedKnownToBeLocal` will be true for `self`. Okay.

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 12]
```diff
@@ -4782,20 +4795,27 @@ VarDecl *swift::getReferencedParamOrCapture(
...
     return false;
 
-  if (var->getName().str().equals("__secretlyKnownToBeLocal")) {
-    // FIXME(distributed): we did a dynamic check and know that this actor is
-    //   local, but we can't express that to the type system; the real
-    //   implementation will have to mark 'self' as "known to be local" after
-    //   an is-local check.
-    return true;
-  }
+//  if (var->getName().str().equals("__secretlyKnownToBeLocal")) {
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:4804]: I suspect you'll need to keep this hack around for a little while.

File: lib/Sema/TypeCheckEffects.cpp [hunk 13]
```diff
@@ -2879,6 +2879,10 @@ class CheckEffectsCoverage : public EffectsHandlingWalker<CheckEffectsCoverage>
            auto callee = call->getCalledValue();
            if (callee) {
              auto declKind = callee->getDescriptiveKind();
+
+             fprintf(stderr, "[%s:%d] (%s) callee:\n", __FILE__, __LINE__, __FUNCTION__);
+             callee->dump();
```
Comment (DougGregor) [inline lib/Sema/TypeCheckEffects.cpp:2884]: Stray debugging output

File: lib/Sema/TypeCheckType.cpp [hunk 14]
```diff
@@ -3676,6 +3690,35 @@ TypeResolver::resolveCompileTimeConstTypeRepr(CompileTimeConstTypeRepr *repr,
...
+      options.hasBase(TypeResolverContext::EnumElementDecl)) {
+    diagnoseInvalid(
+        repr, repr->getSpecifierLoc(), diag::attr_only_on_parameters,
+        "_local");
+    return ErrorType::get(getASTContext());
+  }
+
+  // TODO(distributed): more diagnosis here, prevent from use in props, enums etc
+
+  // isolated parameters must be of 'distributed actor' type
```
Comment (DougGregor) [inline lib/Sema/TypeCheckType.cpp:3709]: s/isolated/_local

File: lib/Sema/TypeCheckType.cpp [hunk 15]
```diff
@@ -3676,6 +3690,35 @@ TypeResolver::resolveCompileTimeConstTypeRepr(CompileTimeConstTypeRepr *repr,
...
+  if (!type->hasTypeParameter() &&
```
Comment (DougGregor) [inline lib/Sema/TypeCheckType.cpp:3710]: Presumably it's okay for it to have a type parameter, but then the archetype would have to be a distributed actor (i.e., conform to the `DistributedActor` protocol).

File: lib/Serialization/ModuleFormat.h [hunk 16]
```diff
@@ -1099,7 +1099,8 @@ namespace decls_block {
     ValueOwnershipField, // inout, shared or owned?
     BCFixed<1>,          // isolated
     BCFixed<1>,          // noDerivative?
-    BCFixed<1>           // compileTimeConst
+    BCFixed<1>,          // compileTimeConst
+    BCFixed<1>           // distributedKnownLocal
```
Comment (DougGregor) [inline lib/Serialization/ModuleFormat.h:1103]: Please bump the module format version when you make changes.

File: stdlib/public/Distributed/DistributedActor.swift [hunk 17]
```diff
@@ -140,8 +140,18 @@ extension DistributedActor {
   /// state.
   ///
   /// When the actor is remote, the closure won't be executed and this function will return nil.
+//  public nonisolated func whenLocal<T: Sendable>(
```
Comment (DougGregor) [inline stdlib/public/Distributed/DistributedActor.swift:143]: I assume you want to remove this?

File: test/Distributed/Runtime/distributed_actor_whenLocal.swift (hunk 2 above)
Comment (DougGregor) [reply test/Distributed/Runtime/distributed_actor_whenLocal.swift:34]: I suspect this is because the inferred known-local bits aren't getting into the solution set, so they don't get applied to the `ParamDecl`s.

File: include/swift/AST/Decl.h [hunk 18]
```diff
@@ -5672,12 +5679,18 @@ class ParamDecl : public VarDecl {
...
+//    /// Whether or not this parameter is '_local'.
+//    IsDistributedKnownLocal = 1 << 3, // FIXME: can't do this because: /llvm-project/llvm/include/llvm/ADT/PointerIntPair.h:148:3: error: static_assert failed due to requirement '4U <= PointerLikeTypeTraits<swift::ParamDecl::StoredDefaultArgument *>::NumLowBitsAvailable' "PointerIntPair with integer size too large for pointer"
   };
 
   /// The default value, if any, along with flags.
   llvm::PointerIntPair<StoredDefaultArgument *, 3, OptionSet<Flags>>
       DefaultValueAndFlags;
 
+  // FIXME: can we pack this somewhere as flags, rather than ad hoc bool?
+  bool isDistributedKnownLocal = false;
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/AST/Decl.h:5692]: Ok cool, I was very iffy about how I packed it into the flags but seems that's fine -- removing the bool then, thanks!

File: lib/Sema/TypeCheckConcurrency.cpp (hunk 11 above)
Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp:2302]: Right, self is a special case that is known local :)

File: lib/Sema/TypeCheckConcurrency.cpp (hunk 12 above)
Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp:4804]: Okey will do 👍

File: stdlib/public/Distributed/DistributedActor.swift (hunk 17 above)
Comment (ktoso) [reply stdlib/public/Distributed/DistributedActor.swift:143]: I have to figure out how to introduce the "good ones" without breaking existing code; this one has unfortunate semantics that it does "await x.whenLocal <HOP HERE> { isolated X in .... }` whereas the other one this `_local` enables allows us to not hop unless we have to: `x.whenLocal { _local X in xx.someConstant }` so there's no hop and no async there...

We also want to allow `await x.whenLocal { _local X in await xx.something }` though; so we can either do this by keeping the old version which does the HOP once (could be good?), or making some new overload.

So in other words... perhaps keeping the "old" signature is good because:
- won't break API/ABI
- is could be okey...? we do one hop when the function is async and run on the actor...?

If we could get away with removing this function though I think we would want to -- can we though?

File: lib/Sema/TypeCheckType.cpp (hunk 15 above)
Comment (ktoso) [reply lib/Sema/TypeCheckType.cpp:3710]: Good point, I think that's right... will think about it and try

File: lib/Sema/TypeCheckConcurrency.cpp (hunk 8 above)
Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp:2136]: Explained below, it is subsumed by checking "is known local?", which `self` just happens to be 👍

File: lib/Sema/CSSimplify.cpp (hunk 6 above)
Comment (ktoso) [reply lib/Sema/CSSimplify.cpp:9872]: Not sure I understood how to take action on this bit -- agreed, but where would I centralize this? 👍

File: lib/AST/Expr.cpp (hunk 4 above)
Comment (ktoso) [reply lib/AST/Expr.cpp:281]: Yes, I'll we'll dig through the expr to find the `base` of the Call or property access, and if it has a `_local` we say `true`.

I always have trouble figuring out how to find the base, but I think we've done this elsewhere before.

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (ktoso) [reply include/swift/AST/Decl.h:5644]: Aha, I see, thank you!

File: include/swift/AST/Decl.h [hunk 19]
```diff
@@ -5672,12 +5679,18 @@ class ParamDecl : public VarDecl {
 
     /// Whether or not this parameter is 'isolated'.
     IsIsolated = 1 << 2,
+
+//    /// Whether or not this parameter is '_local'.
+//    IsDistributedKnownLocal = 1 << 3, // FIXME: can't do this because: /llvm-project/llvm/include/llvm/ADT/PointerIntPair.h:148:3: error: static_assert failed due to requirement '4U <= PointerLikeTypeTraits<swift::ParamDecl::StoredDefaultArgument *>::NumLowBitsAvailable' "PointerIntPair with integer size too large for pointer"
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/AST/Decl.h:5684]: Ok great; wasn't sure if one of the places is more "proper" but I guess it is just about "we have a spare bit here" :)

File: test/Distributed/Runtime/distributed_actor_whenLocal.swift (hunk 2 above)
Comment (ktoso) [reply test/Distributed/Runtime/distributed_actor_whenLocal.swift:34]: Oh okey so that's going to be from the solution... great, I know how to proceed with those, thanks!

File: lib/Sema/CSSimplify.cpp (hunk 7 above)
Comment (ktoso) [reply lib/Sema/CSSimplify.cpp:9873]: Gotcha thanks!

**Comment by ktoso** [comment]
I think I managed to implement this for methods -- properties are hard... I'll need to consult a bit with you @DougGregor but they can also come later I think.

File: lib/Serialization/ModuleFormat.h (hunk 16 above)
Comment (ktoso) [reply lib/Serialization/ModuleFormat.h:1103]: bumped

File: include/swift/AST/Decl.h [hunk 20]
```diff
@@ -5878,6 +5891,25 @@ class ParamDecl : public VarDecl {
                                       : flags - Flags::IsIsolated);
   }
 
+  /// Whether or not this parameter is marked with 'isolated'.
+  bool isDistributedKnownToBeLocal() const {
+    auto yes = getAttrs().hasAttribute<DistributedKnownToBeLocalAttr>() ||
+        ArgumentNameAndFlags.getInt().contains(ArgumentNameFlags::IsDistributedKnownToBeLocal);
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/AST/Decl.h]: Done and cleaned up!

File: lib/Sema/TypeCheckConcurrency.cpp (hunk 12 above)
Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp]: keeping it

File: stdlib/public/Distributed/DistributedActor.swift (hunk 17 above)
Comment (ktoso) [reply stdlib/public/Distributed/DistributedActor.swift]: Open question what to do here, the "I'd love those" signatures are:


```swift
  public nonisolated func whenLocal<T: Sendable>( // FIXME: do we need sendable req here?
    _ body: @Sendable (_local Self) async throws -> T
  ) async rethrows -> T? {
    if __isLocalActor(self) {
       return try await body(self)
    } else {
      return nil
    }
  }

  public nonisolated func whenLocal<T: Sendable>(
    _ body: (_local Self) throws -> T
  ) rethrows -> T? {
    if __isLocalActor(self) {
       return try body(self)
    } else {
      return nil
    }
  }
```


so we can avoid the "one hop" at the beginning, but I guess we COULD keep the isolated one as well hmm

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: include/swift/AST/Decl.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (DougGregor) [reply include/swift/AST/Decl.h]: Yeah, it's conceptually one flag set that's been split because there are spare bits in a few different places.

File: stdlib/public/Distributed/DistributedActor.swift (hunk 17 above)
Comment (DougGregor) [reply stdlib/public/Distributed/DistributedActor.swift]: These signatures look fine to me, but I don't think you need the Sendable constraints because you're never actually crossing the actor boundary. But I think you do need to decide to either have the `_local` versions, or the `isolated` versions, but if you want both you're going to need to give them different names.

**Comment by DougGregor** [comment]
FYI, Windows is telling us there's a real issue here:
```
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\lib\AST\Expr.cpp(282) : error C4716: 'swift::Expr::isDistributedKnownToBeLocal': must return a value
[401/1275] Building CXX object lib\AST\CMakeFiles\swiftAST.dir\RequirementMachine\ConcreteContraction.cpp.obj
```
I don't think we need `Expr::isDistributedKnownToBeLocal` at all, right?

File: stdlib/public/Distributed/DistributedActor.swift [hunk 21]
```diff
@@ -135,20 +135,32 @@ extension DistributedActor {
...
+  /// The `Self` passed to the body closure is guaranteed to be local, and
+  /// accesses on it bypass the distributed isolation checks, meaning that the
   /// closure can be used to call non-distributed functions, or even access actor
-  /// state.
+  /// state using actor-isolation semantics (rather than distributed actor-isolation semantics).
   ///
   /// When the actor is remote, the closure won't be executed and this function will return nil.
-  public nonisolated func whenLocal<T: Sendable>(
-    _ body: @Sendable (isolated Self) async throws -> T
+  public nonisolated func whenLocal<T: Sendable>( // FIXME: do we need sendable req here?
```
Comment (DougGregor) [inline stdlib/public/Distributed/DistributedActor.swift:144]: As noted elsewhere, I don't think we need a Sendable requirements here.

Comment (DougGregor) [reply stdlib/public/Distributed/DistributedActor.swift:144]: Should these be `@_alwaysEmitIntoClient`?

File: lib/Sema/TypeCheckType.cpp (hunk 15 above)
Comment (ktoso) [reply lib/Sema/TypeCheckType.cpp:3710]: Ok awesome :)

File: lib/Sema/TypeCheckAttr.cpp [hunk 22]
```diff
@@ -5915,9 +5915,41 @@ void AttributeChecker::visitDistributedActorAttr(DistributedActorAttr *attr) {
...
+    // TODO(distributed): diagnoseAndRemoveAttr, seems we have wrong position for it though
+    return;
   }
+
+  // We only allow the `da.whenLocal { (...: _local DA) in }` to use '_local',
+  // currently, meaning that the local parameter will always be a closure
+  // param, since we're not able to safely carry it around and propagate the
+  // "local-ness" just yet.
+  // TODO(distributed): unlock this once '_local' is also implemented in properties
+  if (auto closureExpr = dyn_cast<AbstractClosureExpr>(D->getDeclContext())) {
```
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckAttr.cpp:5932]: > Why should _local not be permitted anywhere that isolated is?

Just because I don't think the assignments are handled properly, but maybe it'll just work when implemented in

File: lib/Sema/TypeCheckAttr.cpp [hunk 23]
```diff
@@ -5915,9 +5915,41 @@ void AttributeChecker::visitDistributedActorAttr(DistributedActorAttr *attr) {
 }
 
 void AttributeChecker::visitDistributedKnownToBeLocalAttr(DistributedKnownToBeLocalAttr *attr) {
-  if (!D->isImplicit()) {
+  if (D->isImplicit()) {
+    // Implicit code is fine, this is how we use '_local' in synthesized thunks.
+    return;
```
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckAttr.cpp:5920]: Hmm... I think that means we'll have to change synthesis of thunks to call `whenLocal`, because the thunks do "if local ... call the real func" (without using the whenLocal -- at least today), and they do so from `nonisolated` func -- so we'll have to change thunk synthesis to use whenLocal to get the `_local self` in order to do this 🤔 

That probably makes sense though, so thunks should become to use the normal function from here I suppose 🤔

File: include/swift/AST/Decl.h [hunk 24]
```diff
@@ -5906,6 +5900,10 @@ class ParamDecl : public VarDecl {
 
   void setDistributedKnownToBeLocal(bool value = true) {
     auto flags = ArgumentNameAndFlags.getInt();
+    if (value && !getAttrs().hasAttribute<DistributedKnownToBeLocalAttr>()) {
+      getAttrs().add(new (getASTContext())
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/AST/Decl.h:5892]: Right, thank you!

**Comment by ktoso** [comment]
> I don't think we need Expr::isDistributedKnownToBeLocal at all, right?

Yeah I think you're right, we do it in one centralized spot anyway, will remove this.

File: stdlib/public/Distributed/DistributedActor.swift (hunk 21 above)
Comment (ktoso) [reply stdlib/public/Distributed/DistributedActor.swift:144]: Probably good idea, these impls will never change 👍

File: lib/Sema/TypeCheckAttr.cpp [hunk 25]
```diff
@@ -5915,9 +5915,41 @@ void AttributeChecker::visitDistributedActorAttr(DistributedActorAttr *attr) {
...
+    if (!nominal->isDistributedActor()) {
+      diagnoseAndRemoveAttr(
+          attr, diag::distributed_local_cannot_be_used_with_non_da_type,
+          nominal->getDeclaredType(), nominal->getDescriptiveKind());
+      // TODO(distributed): diagnoseAndRemoveAttr, seems we have wrong position for it though
+      return;
+    }
+  } else {
     diagnoseAndRemoveAttr(attr, diag::distributed_local_cannot_be_used);
+    // TODO(distributed): diagnoseAndRemoveAttr, seems we have wrong position for it though
```
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckAttr.cpp]: Yes, sorry for stray TODO!

File: lib/Parse/ParsePattern.cpp [hunk 26]
```diff
@@ -623,10 +625,17 @@ mapParsedParameters(Parser &parser,
           }
 
           if (auto *STR = dyn_cast<SpecifierTypeRepr>(unwrappedType)) {
-            if (isa<IsolatedTypeRepr>(STR))
+            if (isa<IsolatedTypeRepr>(STR)) {
               param->setIsolated(true);
+            } else if (isa<DistributedKnownToBeLocalTypeRepr>(STR)) {
+              param->setDistributedKnownToBeLocal(true);
+              auto localAttr = new (parser.Context)
```
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Parse/ParsePattern.cpp:632]: Yeah that may be true -- perhaps it is indeed much more like `isolated` than the `const` attribute I based a lot of this work on... I'll try if we can get rid of the attr

**Comment by ktoso** [comment]
This will need to be redone since we're changing back how thunks work.
